src/utils/OpenProXLLF.cpp

```diff
@@ -0,0 +1,1007 @@
+      }
+
+      const double precursor_charge = spectrum.getPrecursors()[0].getCharge();
+      const double precursor_mz = spectrum.getPrecursors()[0].getMZ();
+      const double precursor_mass = (precursor_mz * static_cast<double>(precursor_charge)) - (static_cast<double>(precursor_charge) * Constants::PROTON_MASS_U);
+
+      // needed farther down in the scoring, but only needs to be computed once for a spectrum
+      vector< double > aucorrx = OpenProXLUtils::xCorrelation(spectrum, spectrum, 5, 0.3);
+      vector< double > aucorrc = OpenProXLUtils::xCorrelation(spectrum, spectrum, 5, 0.2);
+
+      vector< CrossLinkSpectrumMatch > top_csms_spectrum;
+
+      // determine candidates
+      vector< OpenProXLUtils::XLPrecursor > candidates;
+      double allowed_error = 0;
+
+      // determine MS2 precursors that match to the current peptide mass
+      vector< OpenProXLUtils::XLPrecursor >::const_iterator low_it;
+      vector< OpenProXLUtils::XLPrecursor >::const_iterator up_it;
+
+      if (precursor_mass_tolerance_unit_ppm) // ppm
+      {
+        allowed_error = precursor_mass * precursor_mass_tolerance * 1e-6;
+      }
+      else // Dalton
+      {
+        allowed_error = precursor_mass_tolerance;
+      }
+#ifdef _OPENMP
+#pragma omp critical (enumerated_cross_link_masses_access)
+#endif
+      {
+        low_it = lower_bound(enumerated_cross_link_masses.begin(), enumerated_cross_link_masses.end(), precursor_mass - allowed_error);
+        up_it =  upper_bound(enumerated_cross_link_masses.begin(), enumerated_cross_link_masses.end(), precursor_mass + allowed_error);
+      }
+
+      if (low_it != up_it) // no matching precursor in data
+      {
+        for (; low_it != up_it; ++low_it)
+        {
+          candidates.push_back(*low_it);
+        }
+      }
+
+#ifdef _OPENMP
+#pragma omp critical
+#endif
+      cout << "Number of candidates for this spectrum: " << candidates.size() << endl;
+
+      // Find all positions of lysine (K) in the peptides (possible scross-linking sites), create cross_link_candidates with all combinations
+      vector <ProteinProteinCrossLink> cross_link_candidates = OpenProXLUtils::buildCandidates(candidates, peptide_masses, cross_link_residue1, cross_link_residue2, cross_link_mass, cross_link_mass_mono_link, precursor_mass, allowed_error, cross_link_name, n_term_linker, c_term_linker);
+
+      // lists for one spectrum, to determine best match to the spectrum
+      vector< CrossLinkSpectrumMatch > all_csms_spectrum;
+
+#ifdef _OPENMP
+#pragma omp critical (max_subscore_variable_access)
+#endif
+      {
+        if (cross_link_candidates.size() > maxMatchCount) maxMatchCount = cross_link_candidates.size();
+        sumMatchCount += cross_link_candidates.size();
+      }
+
+      for (Size i = 0; i != cross_link_candidates.size(); ++i)
+      {
+        ProteinProteinCrossLink cross_link_candidate = cross_link_candidates[i];
```

Comment:
it looks like a large part of this code is duplicated as well between the two tools